# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [16]:
# Your code here

# set-up, load the dataset

# dataset location:  https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

import pandas as pd

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
data = pd.read_csv(url)
print(data.head())


  Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car 

In [17]:
# Identify the dimensions of the dataset 
# by determining the number of rows and columns it contains.
print("Dataset dimensions:")
print(data.shape)

Dataset dimensions:
(4008, 11)


> **Conclusion Dataset Dimensions:**
>
> The dataset contains 4,008 rows and 11 columns, which represents a medium-sized dataset.
>
> From a purely quantitative perspective, this size is appropriate for exploratory data analysis using pandas and does not present performance limitations.
>
> In principle, the number of rows appears sufficient to extract meaningful statistical insights and identify patterns. However, the reliability of these insights strongly depends on data quality. Issues such as inconsistent categorical values, missing data, incorrect data types, or outliers may reduce the validity of the results. Therefore, data cleaning and validation are necessary steps before drawing definitive conclusions or using the data for further analysis or modeling.

In [18]:
# Determine the data types of each column 
# and evaluate whether they are appropriate for the nature of the variable. 

print("\nData types of each column:")
print(data.dtypes)



Data types of each column:
Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object


> **Datatyppe Evaluation:**
> 
> Based on the output of data.dtypes, most columns have data types that generally align with the nature of the variables. However, several important issues were identified.
>
> **1. Appropriate data types:**
>
> **Customer (object):**
 The data type is appropriate, as this column represents a unique customer identifier and should be treated as a categorical variable rather than a numerical one.
>
> **ST (object):**
 The data type is appropriate because this column represents categorical information (customer state).
>
> **GENDER (object):**
 The data type is appropriate for a categorical variable and gender is a categorical variable.
>
> **Education (object):**
 The data type is appropriate since education level is a categorical variable.
>
> **Policy Type (object):**
 The data type is appropriate because it represents distinct policy categories.
>
> **Vehicle Class (object):**
 The data type is appropriate, as this variable represents different vehicle categories.
>
> **Income (float64):**
 The data type is appropriate for a continuous numerical variable and Income is a continous numerical variable.
>
> **Monthly Premium Auto (float64):**
 The data type is appropriate for a numerical variable representing a monetary value.
>
> **Total Claim Amount (float64):**
 The data type is appropriate for a numerical variable representing monetary amounts.
>
>
> **2. Data types requiring correction or clarification:**
>
> **Customer Lifetime Value (object):**
 This data type is not appropriate for the nature of the variable. Customer Lifetime Value is a numerical measure, but it is stored as an object due to the presence of percentage symbols (%) as we could see in our first output when we seted up and loaded the dataset.
>
> *Recommendation:* Remove the % symbol and convert this column to a numerical type (float) to enable proper quantitative analysis.
>
> **Number of Open Complaints (object):**
 This data type is potentially inappropriate. Although stored as an object, the values appear to encode numerical counts using a date-like format (e.g., 1/0/00).
 >
 > *Recommendation:* Clarify the semantic meaning of this column and convert it to an integer representing the number of complaints.
>
>
> **Overall assessment:**
> In summary, most variables are stored using appropriate data types. However, two columns require transformation to ensure correct interpretation and analysis. Addressing these issues is essential to improve data quality and avoid misleading analytical results.
>
> **Categorical variables** (Customer, ST, GENDER, Education, Policy Type, Vehicle Class)
 → Are appropiarly stored as object in the DataFrame.
>
> **Numerical variables** (Income, Monthly Premium Auto, Total Claim Amount)
 → Are also correctly stored as float64.
>
> **Problematic variables:**
>
> *Customer Lifetime Value:* this variable should be numerical, but it is stored as object. This need to be converted to a float.
>
> *Number of Open Complaints:* this variable represents a discrete numerical count, but it is stored as object. This need to be converted to an interger.

In [19]:
# Identify the number of unique values for each column 
# and determine which columns appear to be categorical. 
# You should also describe the unique values of each categorical column 
# and the range of values for numerical columns, and give your insights.

print("\nUnique values and ranges for each column:")
for column in data.columns:
    unique_values = data[column].nunique()
    print(f"\nColumn: {column}")
    print(f"Number of unique values: {unique_values}")
    if data[column].dtype == 'object':
        print("Categorical values:")
        print(data[column].value_counts())
    else:
        print("Numerical range:")
        print(f"Min: {data[column].min()}, Max: {data[column].max()}")


Unique values and ranges for each column:

Column: Customer
Number of unique values: 1071
Categorical values:
Customer
MY31220    1
RB50392    1
QZ44356    1
AI49188    1
WW63253    1
          ..
FL50705    1
WC83389    1
OE15005    1
FV94802    1
BW63560    1
Name: count, Length: 1071, dtype: int64

Column: ST
Number of unique values: 8
Categorical values:
ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

Column: GENDER
Number of unique values: 5
Categorical values:
GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64

Column: Education
Number of unique values: 6
Categorical values:
Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64

Column: Customer Lifetime Va

> **Evaluation of Value Consistency and Data Quality:**
>
> This section evaluates the consistency of values across all variables based on their unique values and distributions. For each variable, potential inconsistencies are identified, their impact on data quality is assessed, and recommendations are provided where necessary.
>
> **1. Customer:**
>
> *Evaluation of consistency*
> The column contains 1,071 unique values, each appearing exactly once.
> This behavior is consistent with a unique customer identifier.
>
> *Impact on data quality*
> No negative impact. The variable behaves as expected.
>
> *Recommendation*
> No action required. The column should be treated as an identifier and excluded from aggregations or statistical analysis.
>
>
> **2. State (ST):**
>
> *Evaluation of consistency*
> The column contains 8 unique values.
> 
> Multiple representations exist for the same state:
> - California vs Cali
> - Washington vs WA
> - Arizona vs AZ
>
> *Inconsistency identified*
> Semantic duplication caused by inconsistent naming conventions.
> 
> *Impact on data quality*
> - This fragmentation leads to incorrect aggregations and misleading geographic insights.
> - State-level analyses may underestimate or overestimate customer counts.
>
> *Recommendation*
> - Standardize state names using a consistent format (e.g. full state names).
> - Apply a mapping dictionary to normalize values.
>
>
> **3. Gender:**
>
> *Evaluation of consistency*
> The column contains 5 unique values representing only two genders.
> 
> Inconsistencies include:
> - Abbreviations (F, M)
> - Full words (Male, female)
> - Typographical errors (Femal)
> - Inconsistency identified
> - Mixed formats and spelling errors.
>
> *Impact on data quality*
> - Gender-based analysis becomes unreliable.
> - Grouping operations may produce fragmented or incorrect results.
>
> *Recommendation*
> - Normalize values into a standardized set (e.g. Male, Female).
> - Optionally handle missing or ambiguous values explicitly.
>
>
> **4. Education:**
>
> *Evaluation of consistency*
> The column contains 6 unique values.
>
> Minor duplication exists:
> - Bachelor vs Bachelors
>
> *Inconsistency identified*
> - Semantic duplication of equivalent education levels.
>
> *Impact on data quality*
> - Slight distortion in education-level distributions.
> - Reduced clarity for segmentation analysis.
>
> *Recommendation*
> - Merge equivalent categories into a single standardized value.
> - Optionally define an ordinal hierarchy if used for modeling.
>
>
> **5. Customer Lifetime Value:**
>
> *Evaluation of consistency*
> - The column contains 1,027 unique values, indicating a continuous variable.
> - Values are stored as strings with a percentage symbol (%).
>
> *Inconsistency identified*
> - Numerical values stored as text.
> - Misclassification as categorical data.
> 
> *Impact on data quality*
> - Prevents numerical analysis, aggregation, and modeling.
> - Increases risk of misinterpretation.
>
> *Recommendation*
> - Remove the % symbol and convert the column to a numerical type (float).
> - Treat this variable as continuous rather than categorical.
>
>
> **6. Income:**
>
> *Evaluation of consistency*
> Contains 774 unique values with a range from 0 to 99,960.
> Distribution appears plausible for income data.
>
> *Potential issue identified*
> - A significant number of zero values may represent missing or default entries.
>
> *Impact on data quality*
> - ero values may bias average income calculations.
>
> *Recommendation*
> - Investigate whether zero income represents missing data.
> - Consider replacing invalid zeros with NaN if appropriate.
>
>
> **7. Monthly Premium Auto**
>
> *Evaluation of consistency*
> Contains 132 unique values.
> Range from 61 to 35,354.
>
> *Potential issue identified*
> - Presence of extreme values suggests outliers.
>
> *Impact on data quality*
> Outliers may skew averages and standard deviation.
> 
> *Recommendation*
> - Investigate unusually high premiums.
> - Consider outlier treatment (capping or filtering) depending on analysis goals.
>
>
> **8. Number of Open Complaints:**
>
> *Evaluation of consistency*
> Contains 6 unique values, all encoded in a date-like format (1/0/00, 1/1/00, etc.).
>
> *Inconsistency identified*
> Complaint counts are stored using an ambiguous and misleading format.
>
> *Impact on data quality*
> - High risk of misinterpretation.
> - Limits usability for numerical analysis.
>
> *Recommendation*
> - Extract the numerical complaint count.
> - Convert the column to an integer type representing the number of complaints.
>
> 
> **9. Policy Type:**
>
> *Evaluation of consistency*
> Contains 3 unique values.
> Categories are clearly defined and consistently labeled.
> 
> *Impact on data quality*
> No issues identified.
>
> *Recommendation*
> No action required.
>
> **10. Vehicle Class:**
>
> *Evaluation of consistency*
> Contains 6 unique values.
> Categories are well defined and internally consistent.
>
> *Impact on data quality*
> No issues identified.
>
> *Recommendation*
> No action required.
>
>
> **11. Total Claim Amount**
>
> *Evaluation of consistency*
> Contains 761 unique values.
> Range appears realistic for insurance claim amounts.
> 
> *Impact on data quality*
> No structural issues identified.
>
> *Recommendation*
> - Variable is suitable for numerical analysis.
> - Consider outlier analysis in later stages.
>
> 
> **Overall Assessment:**
>
> While the dataset contains a sufficient number of observations, several value consistency issues were identified, primarily affecting categorical variables and incorrectly typed numerical fields. These issues can significantly reduce the reliability of aggregated insights and analytical results. Addressing these inconsistencies through data cleaning and standardization is a necessary step before performing advanced analysis or modeling.

In [20]:
#  Compute summary statistics such as mean, median, mode, standard deviation, 
# and quartiles to understand the central tendency 
# and distribution of the data for numerical columns. 
# You should also provide your conclusions based on these summary statistics.
print("\nSummary statistics for numerical columns:")
print(data.describe())


Summary statistics for numerical columns:
             Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678


> **Conclusions Based on Summary Statistics:**
>
> The summary statistics provide insights into the central tendency, dispersion, and distribution of the numerical variables in the dataset: Income, Monthly Premium Auto, and Total Claim Amount.
>
> 
> **1. Income:**
> - The mean income is approximately 39,295.70, while the median is 36,234.00.
> The mean being higher than the median suggests a right-skewed distribution, where higher-income values pull the average upward.
> - The minimum income is 0, indicating that some customers report no income.
> The wide range (0 to 99,960) and high standard deviation (30,469.43) suggest substantial variability in customer income levels.
>
> *Conclusion:*
> Income is unevenly distributed across customers, with significant dispersion and potential outliers. The presence of zero-income values may indicate missing data or special customer categories, which could bias income-based analyses if not addressed.
>
>
> **2. Monthly Premium Auto:**
> - The median monthly premium is 83.00, while the mean is considerably higher at 193.23.
> This large difference between the mean and the median indicates a strong right-skewed distribution.
> - The maximum value (35,354) is extremely high relative to the upper quartile (109.50), suggesting the presence of extreme outliers.
> The standard deviation (1,601.19) is disproportionately large compared to the interquartile range.
>
> *Conclusion:*
Most customers pay relatively low monthly premiums, while a small number of extreme values heavily distort the average. Outlier analysis is strongly recommended before using this variable for modeling or business decisions.
>
> 
> **3. Total Claim Amount:**
> - The mean claim amount is approximately 404.99, while the median is 354.73.
> The distribution is moderately right-skewed, with higher claim values affecting the mean.
> - The interquartile range (approximately 202.16 to 532.80) shows that most customers fall within a relatively narrow claim range.
> The maximum value (2,893.24) indicates the presence of high-claim customers.
>
> *Conclusion:*
Most customers generate relatively low to moderate claim amounts, while a smaller group accounts for significantly higher claims. This variable is well suited for identifying high-risk or high-cost customer segments.
>
> 
> **Overall Conclusions**
> - All numerical variables exhibit right-skewed distributions, indicating the presence of outliers.
> - The median provides a more representative measure of central tendency than the mean for all variables.
> - High variability in income and premium values suggests heterogeneous customer profiles.
> - Before conducting advanced analysis or predictive modeling, outlier treatment and data validation are recommended to improve the robustness of insights.


In [21]:
# Compute summary statistics for categorical columns 
# and providing your conclusions based on these statistics.

print("\nSummary statistics for categorical columns:")
for column in data.select_dtypes(include=['object']).columns:
    print(f"\nColumn: {column}")
    print(data[column].describe())



Summary statistics for categorical columns:

Column: Customer
count        1071
unique       1071
top       MY31220
freq            1
Name: Customer, dtype: object

Column: ST
count       1071
unique         8
top       Oregon
freq         320
Name: ST, dtype: object

Column: GENDER
count     954
unique      5
top         F
freq      457
Name: GENDER, dtype: object

Column: Education
count         1071
unique           6
top       Bachelor
freq           324
Name: Education, dtype: object

Column: Customer Lifetime Value
count           1068
unique          1027
top       251459.20%
freq               4
Name: Customer Lifetime Value, dtype: object

Column: Number of Open Complaints
count       1071
unique         6
top       1/0/00
freq         830
Name: Number of Open Complaints, dtype: object

Column: Policy Type
count              1071
unique                3
top       Personal Auto
freq                780
Name: Policy Type, dtype: object

Column: Vehicle Class
count              1

> **Conclusions Based on Summary Statistics for Categorical Variables:**
>
> The summary statistics for the categorical variables provide insights into data completeness, category dominance, and potential data quality issues.
>
> 
> **1. Customer:**
> - The column contains 1,071 non-null values, all of which are unique.
> - The most frequent value appears only once.
>
> *Conclusion:*
> This column functions purely as a unique identifier. It does not provide analytical value for aggregation or statistical analysis and should be excluded from modeling and descriptive summaries.
>
> 
> **2. State (ST):**
> - The column contains 1,071 non-null values and 8 unique categories.
> - Oregon is the most frequent state, with 320 occurrences.
>
> *Conclusion:*
> State distribution is uneven, with a clear dominance of certain locations. However, the presence of multiple representations for the same state (e.g., abbreviations and full names) indicates inconsistent categorical encoding, which may distort geographic analyses.
>
> 
> **3. Gender:**
> - The column contains 954 non-null values, indicating missing data.
> - There are 5 unique categories, with F being the most frequent value.
>
> *Conclusion:*
> Gender data is incomplete and inconsistently encoded. The presence of multiple labels for the same gender reduces reliability and complicates demographic analysis. Standardization and missing value handling are necessary before further use.
>
>
> **4. Education:**
> - The column contains 1,071 non-null values and 6 unique categories.
> - Bachelor is the most common education level.
> 
> *Conclusion:*
> Education data is largely complete and informative, but minor inconsistencies (e.g., Bachelor vs Bachelors) may slightly distort category distributions. Consolidation of equivalent categories is recommended.
>
>
> **5. Customer Lifetime Value:**
> - The column contains 1,068 non-null values, indicating some missing values.
> - There are 1,027 unique values, which is unusually high for a categorical variable.
>
> *Conclusion:*
> Although currently treated as categorical, this variable behaves like a continuous numerical measure. Storing it as text prevents meaningful statistical analysis. Conversion to a numerical format is essential to unlock its analytical value.
>
>
> **6. Number of Open Complaints**
> - The column contains 1,071 non-null values and 6 unique values.
> - The most frequent value (1/0/00) appears 830 times.
>
> *Conclusion:*
> This variable shows a highly skewed distribution, with most customers reporting zero complaints. However, the date-like encoding is misleading and may cause misinterpretation. Re-encoding this variable as an integer count is strongly recommended.
>
> 
> **7. Policy Type:**
> - The column contains 1,071 non-null values and 3 unique categories.
> - Personal Auto is the dominant policy type.
> 
> *Conclusion:*
> Policy type data is complete and consistently encoded. However, the dataset is heavily skewed toward Personal Auto policies, which may bias analyses and conclusions.
>
>
> **8. Vehicle Class:**
> - The column contains 1,071 non-null values and 6 unique categories.
> - Four-Door Car is the most frequent vehicle class.
>
> *Conclusion:*
> Vehicle class data is complete and well structured. The distribution reflects realistic customer preferences and is suitable for segmentation and risk analysis.
>
>
> **Overall Conclusions for Categorical Variables:**
> - Most categorical variables are complete, but several suffer from inconsistent encoding.
> - Identifier variables should be excluded from analysis.
> - Some variables currently treated as categorical should be reclassified or transformed.
> - Data cleaning and standardization are necessary steps to improve analytical reliability.

This exploratory analysis focused on understanding the structure, distributions, and quality of the data. Several inconsistencies and data type issues were identified; however, no transformations were applied at this stage, as the objective of this challenge is limited to analysis and recommendations. The identified issues provide clear guidance for future data cleaning and preparation steps.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [32]:
# Your code here

# The marketing team wants to know the top 5 less common customer locations.
# Create a pandas Series object that contains the customer locations and their frequencies
# sorted in ascending order. Display the top 5 less common customer locations.

customer_location_counts = data['ST'].value_counts(ascending=True)
top_5_less_common_locations = customer_location_counts.head(5)
print("\nTop 5 less common customer locations:")
print(top_5_less_common_locations)



Top 5 less common customer locations:
ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


> **Note on data consistency and interpretation:**
>
>
> The results above are derived directly from the raw dataset, without applying any data cleaning or standardization to the ST (state) variable. As identified in the exploratory analysis, this column contains inconsistencies such as multiple representations of the same state (e.g., abbreviations and full names).
>
> At this stage, the analysis intentionally uses the data as is, in accordance with the objective of this exercise.
>
>
> *Evaluation and potential risk*
>
> While the result correctly identifies the five least common customer locations based on the raw data, these inconsistencies may lead to fragmented counts and misleading geographic insights. States represented under different labels may appear less frequent than they actually are when aggregated correctly.
Therefore, conclusions drawn from this result should be interpreted with caution. For accurate location-based analysis and decision-making, value standardization should be applied in a subsequent data cleaning phase.


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [34]:
# Your code here

# The sales team wants to know the total number of policies sold for each type of policy.
# Create a pandas Series object that contains the policy types and their total number of policies sold
# and then retrieve the policy type with the highest number of policies sold.
# Using value_counts() method simplifies this analysis
# Futhermore, there is a method that returns the index of the maximum value in a column or row

policy_type_counts = data['Policy Type'].value_counts()
most_sold_policy_type = policy_type_counts.idxmax()
print("\nTotal number of policies sold for each type of policy:")
print(policy_type_counts)
print(f"\nPolicy type with the highest number of policies sold: {most_sold_policy_type}")



Total number of policies sold for each type of policy:
Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

Policy type with the highest number of policies sold: Personal Auto


> **Note:**
>
> The results show that Personal Auto is by far the most common policy type, accounting for the majority of policies sold. Corporate Auto and Special Auto policies represent a significantly smaller portion of the dataset.
>
> Since the Policy Type variable is consistently encoded and does not present data quality issues, these results can be considered reliable. However, the strong dominance of Personal Auto policies indicates an imbalanced distribution, which should be taken into account when performing comparative analyses across policy types.


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [38]:
# Your code here

# The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto.
# How does the average income compare between the two policy types?
# Use loc to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
# Calculate the average income for each policy.
# Print the results.

personal_auto = data.loc[data['Policy Type'] == 'Personal Auto']
corporate_auto = data.loc[data['Policy Type'] == 'Corporate Auto']
avg_income_personal = personal_auto['Income'].mean()
avg_income_corporate = corporate_auto['Income'].mean()
print(f"\nAverage income for Personal Auto customers: {avg_income_personal}")
print(f"Average income for Corporate Auto customers: {avg_income_corporate}")



Average income for Personal Auto customers: 38180.69871794872
Average income for Corporate Auto customers: 41390.31196581197


> **Note:**
>
> The results indicate that customers with Corporate Auto policies have a higher average income than those with Personal Auto policies. This suggests that corporate policyholders tend to belong to higher-income segments. However, this comparison is based on raw data and does not account for potential outliers or zero-income values, which may influence the average.

### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [39]:
# Your code here

# Your goal is to identify customers with a high policy claim amount.
# Review again the statistics for total claim amount to gain an understanding of the data.
# To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount.
# Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount.
# Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
# Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 
# Use method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.

quantile_75 = data['Total Claim Amount'].quantile(0.75)
high_claims = data.loc[data['Total Claim Amount'] > quantile_75]
print("\nCustomers with high policy claim amounts (greater than 75th percentile):")
print(high_claims)
print("\nSummary statistics for high policy claim amounts:")
print(high_claims.describe())



Customers with high policy claim amounts (greater than 75th percentile):
     Customer          ST GENDER Education Customer Lifetime Value   Income  \
1     QZ44356     Arizona      F  Bachelor              697953.59%      0.0   
2     AI49188      Nevada      F  Bachelor             1288743.17%  48767.0   
17    OE15005        Cali    NaN   College              394524.16%  28855.0   
23    TZ98966      Nevada    NaN  Bachelor              245019.10%      0.0   
26    US89481  California    NaN  Bachelor              394637.21%      0.0   
...       ...         ...    ...       ...                     ...      ...   
1059  YG44474      Oregon      M   College             1401472.13%  54193.0   
1061  RY92647        Cali      F  Bachelor             1050677.17%      0.0   
1068  GS98873     Arizona      F  Bachelor              323912.47%  16061.0   
1069  CW49887  California      F    Master              462680.11%  79487.0   
1070  MY31220  California      F   College              8

> **Note:**
>
> The analysis above is based on the raw dataset. Previously identified data quality issues (e.g., inconsistent categorical values and numerical fields stored as text) have not been addressed at this stage. While the results correctly identify the top 25% of customers by claim amount, further data cleaning may be required before drawing operational or strategic conclusions.